staging/src/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/writers.go

```diff
@@ -245,11 +274,16 @@ func (w *deferredResponseWriter) Write(p []byte) (n int, err error) {
 	return w.w.Write(p)
 }
 
-func (w *deferredResponseWriter) Close() error {
+func (w *deferredResponseWriter) Close() (err error) {
 	if !w.hasWritten {
-		return nil
+		if !w.hasBuffered {
```

Comment:
I don't think we should fast return here.
If it is direct write with gzip, we should release gzip.Writer. for example, `p []byte` > 128 K